--- build_dir/target-mips_24kc_musl/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js	2022-03-25 15:40:50.067725797 +0100
+++ build_dir/target-mips_24kc_musl/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js	2022-03-29 13:14:09.794885510 +0200
@@ -321,7 +321,7 @@
 			for (var i = 0; i < data[1].length; i++) {
 				var band;
 
-				if (data[1][i].mhz >= 2412 && data[1][i].mhz <= 2484)
+				if (data[1][i].mhz >= 21 && data[1][i].mhz <= 2484)
 					band = '2g';
 				else if (data[1][i].mhz >= 5160 && data[1][i].mhz <= 5885)
 					band = '5g';
